streampipes-service-core/src/main/java/org/apache/streampipes/service/core/migrations/v095/MergeFilenamesAndRenameDuplicatesMigration.java

```diff
@@ -0,0 +1,217 @@
+
+  protected static final String ORIGINAL_FILENAME = "originalFilename";
+  protected static final String INTERNAL_FILENAME = "internalFilename";
+  protected static final String ID = "_id";
+  protected static final String FILETYPE = "filetype";
+
+  private CouchDbClient couchDbClient;
+
+  private ObjectMapper mapper = new ObjectMapper();
+
+  private IFileMetadataStorage fileMetadataStorage =
+      StorageDispatcher.INSTANCE.getNoSqlStore().getFileMetadataStorage();
+
+  private FileHandler fileHandler = new FileHandler();
+
+  protected Map<String, List<FileMetadata>> fileMetadataGroupedByOriginalName = new HashMap<>();
+
+  private boolean isTesting = false;
+
+  public MergeFilenamesAndRenameDuplicatesMigration(boolean testing) {
+    isTesting = testing;
+  }
+
+  public MergeFilenamesAndRenameDuplicatesMigration() {
+    couchDbClient = Utils.getCouchDbFileMetadataClient();
+  }
+
+  // Starting from v0.95, StreamPipes will use a single file name as the unique identifier of files instead of an
+  // internal filename and an original filename. This migration merges them and renames all the files that have
+  // duplicate names to ensure uniqueness
+  @Override
+  public boolean shouldExecute() {
+    return true;
+  }
+
+  @Override
+  public void executeMigration() {
+    var couchDbRawFileMetadata = getCouchDbRawFileMetadata(getAllFileIds(fileMetadataStorage));
+    getFileMetadataToUpdate(couchDbRawFileMetadata);
+    fileMetadataGroupedByOriginalName.forEach(
+        (originalFilename, fileMetadataList) -> update(originalFilename, fileMetadataList));
+  }
+
+  /**
+   * Gets all fileMetadata that need to be updated grouped by originalFilename
+   * key is (possibly) duplicated originalFilename and value is that file's FileMetadata list (if duplicated)
+   */
+  protected void getFileMetadataToUpdate(List<Map<String, Object>> couchDbRawFileMetadata) {
+    couchDbRawFileMetadata.forEach(
+        rawFileMetadata -> checkDuplicateOriginalFilename(rawFileMetadata));
+  }
+
+  /**
+   * Fetches all fileIds stored in CouchDB
+   */
+  private List<String> getAllFileIds(IFileMetadataStorage fileMetadataStorage) {
+    return fileMetadataStorage.getAllFileMetadataDescriptions().stream().map(fileMetadata -> fileMetadata.getFileId())
+        .toList();
+  }
+
+  /**
+   * Takes the list of fileIds and searches for their raw metadata in CouchDB and returns them
+   */
+  private List<Map<String, Object>> getCouchDbRawFileMetadata(List<String> fileIds) {
+    return fileIds.stream()
+        .map(fileId -> convertInputStreamToMap(couchDbClient.find(fileId)))
+        .toList();
+  }
+
+  /**
+   * Converts InputStream (as stored in CouchDB) to Map, if there's an error, constructs a new Map
+   */
+  private Map<String, Object> convertInputStreamToMap(InputStream inputStream) {
+    try {
+      return mapper.readValue(inputStream, Map.class);
+    } catch (Exception e) {
+      return new HashMap<>();
```

Comment:
Can we please add a log message here?
This allows for better debugging in case we have some problems with the migration

Reply:
Just added a log message. By the way even if this exception occurs, rest of the files' migration will not be affected since it will just be skipped.